org.eclipse.jdt.core.compiler.batch/grammar/java.g

```diff
@@ -1269,30 +1273,48 @@ TypePattern ::= Modifiersopt Type 'Identifier'
 -- 20 preview feature : record patterns
 -----------------------------------------------
 
-RecordPattern ::= Modifiersopt ReferenceType PushLPAREN PatternListopt PushRPAREN
+RecordPattern ::= Modifiersopt ReferenceType PushLPAREN ComponentPatternListopt PushRPAREN
 /.$putCase consumeRecordPattern(); $break ./
 /:$readableName RecordPattern:/
 /:$compliance 20:/
 
-PatternListopt ::=  $empty
+ComponentPatternListopt ::=  $empty
 /.$putCase consumePatternListopt(); $break ./
 /:$readableName PatternListopt:/
 /:$compliance 20:/
 
-PatternListopt -> PatternList
+ComponentPatternListopt -> ComponentPatternList
 /:$readableName PatternListopt:/
 /:$compliance 20:/
 
-PatternList -> Pattern
-PatternList ::= PatternList ',' Pattern
+ComponentPatternList -> ComponentPattern
+ComponentPatternList ::= ComponentPatternList ',' ComponentPattern
 /.$putCase consumePatternList();  $break ./
-/:$readableName PatternList:/
+/:$readableName ComponentPatternList:/
 /:$compliance 20:/
 
+ComponentPattern -> Pattern
+
 -----------------------------------------------
 -- 20 preview feature : end of record patterns
 -----------------------------------------------
 
+---
+TypePattern ::= Modifiersopt Type '_'
+/.$putCase consumeTypePattern(); $break ./
+/:$readableName TypePattern:/
+/:$compliance 21:/
+
+ComponentPattern -> UnnamedPattern
+/:$compliance 21:/
+
+UnnamedPattern ::= '_'
+/.$putCase consumeUnnamedPattern(); $break ./
+/:$readableName UnnamedPattern:/
```

Comment:
Check the uses of `org.eclipse.jdt.internal.compiler.parser.Scanner.Goal.PatternRules` to see if UnnamedPattern must be added there

Reply:
I don't think it needs to be added to the list of PatternRules, since you can't use an unnamed pattern by itself as a case label, and if I understand correctly, `PatternRules` is for all the possible starting states of parsing a case label with a pattern in it.